## Part 2: Statistical Analysis
1. Normality Test (10 points)
    - Perform a normality test (e.g., Shapiro-Wilk test) on the numerical variable you analyzed.
    - Report the test statistic and p-value. Based on these results, do you reject or fail to reject the null hypothesis of normality? Explain your reasoning.
2. Correlation and Covariance (10 points)
    - Calculate the correlation coefficient and covariance between the two numerical variables you used in the scatter diagram.
    - Interpret the results. What do they tell you about the relationship between these variables?


In [1]:
# Load in the dataset and pre-process it just like in part one

import pandas as pd

df = pd.read_csv(
    "household_power_consumption.txt", 
    sep=';', 
    na_values='?', 
    low_memory=False
)

In [2]:
df.head()

Date      Time  Global_active_power  Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00                4.216                  0.418   234.84   
1  16/12/2006  17:25:00                5.360                  0.436   233.63   
2  16/12/2006  17:26:00                5.374                  0.498   233.29   
3  16/12/2006  17:27:00                5.388                  0.502   233.74   
4  16/12/2006  17:28:00                3.666                  0.528   235.68   

   Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
0              18.4             0.0             1.0            17.0  
1              23.0             0.0             1.0            16.0  
2              23.0             0.0             2.0            17.0  
3              23.0             0.0             1.0            17.0  
4              15.8             0.0             1.0            17.0

In [3]:
df.dropna(inplace=True)

In [4]:
df.isnull().sum()

Date                     0
Time                     0
Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
dtype: int64

In [5]:
# Direct computation of normality using Shapiro-Wilk test on numeric varaible `voltage`

from scipy.stats import shapiro

stat, p_value = shapiro(df['Voltage'])

print('Statistic:', stat)
print('p-value:', p_value)


Statistic: 0.9910188290999363
p-value: 3.118078194355427e-91


/Users/sleepdeprived/Documents/Nayan/MTech_AI/Stats/Assignments/jupyter_notebook/.venv/lib/python3.11/site-packages/scipy/stats/_axis_nan_policy.py:592: UserWarning: scipy.stats.shapiro: For N > 5000, computed p-value may not be accurate. Current N is 2049280.
  res = hypotest_fun_out(*samples, **kwds)


The sample size is large for the normality test 
The Shapiro–Wilk test:
- Was designed for small to medium samples
- Uses approximations for large 𝑁
- Becomes unreliable for very large datasets

In [6]:
sample_df = df.sample(n=50, random_state=42)

In [7]:
stat, p_value = shapiro(sample_df['Voltage'])
print('Statistic:', stat)
print('p-value:', p_value)

Statistic: 0.9812151334081245
p-value: 0.6042254088371911


**Conclusion**

A Shapiro–Wilk test was conducted to assess the normality of the Voltage variable using a random sample of 50 observations.
The test produced a statistic of 0.981 and a p-value of 0.604. Since the test statistics value is close to 1 and p-value exceeds the 0.05 significance level, the null hypothesis of normality cannot be rejected, indicating that the Voltage data is approximately normally distributed.

In [8]:
# Mathematical Computation with steps

test_data = (sample_df['Voltage'].sort_values()).to_list()
test_data

[229.35,
 234.34,
 235.36,
 235.53,
 235.72,
 235.91,
 236.79,
 237.02,
 237.45,
 237.54,
 238.17,
 238.36,
 238.44,
 238.53,
 239.05,
 239.1,
 239.65,
 239.77,
 239.85,
 240.1,
 240.17,
 240.17,
 240.27,
 240.54,
 241.01,
 241.13,
 241.34,
 241.45,
 241.54,
 241.67,
 241.94,
 242.2,
 242.32,
 242.65,
 242.78,
 242.97,
 243.17,
 243.24,
 243.29,
 243.9,
 243.91,
 243.95,
 244.36,
 244.52,
 244.82,
 244.92,
 245.22,
 245.5,
 246.63,
 249.24]

In [9]:
sample_mean = sum(test_data)/len(test_data)
sample_mean

240.737

In [10]:
# From Shapiro-wilk column

a_coefficient=[
    0.3751, 0.2574, 0.2260, 0.2032, 0.1847,
    0.1691, 0.1554, 0.1430, 0.1317, 0.1212,
    0.1113, 0.1020, 0.0932, 0.0846, 0.0764,
    0.0685, 0.0608, 0.0532, 0.0459, 0.0386,
    0.0314, 0.0244, 0.0174, 0.0104, 0.0035
]

In [11]:
len(test_data), len(a_coefficient)

(50, 25)

In [12]:
# Calculation of test statistics (W)
# W = b^2/ss

N = len(test_data)
n = len(a_coefficient)
b = 0

# Print Table Header
print(f"{'i':<4} | {'coeff (a_i)':<12} | {'High (x_n-i+1)':<15} | {'Low (x_i)':<10} | {'Difference':<12} | {'Product':<10} | {'Running b'}")
print("-" * 90)

for i in range(n):
    high_val = test_data[N - 1 - i]
    low_val = test_data[i]
    diff = high_val - low_val
    product = a_coefficient[i] * diff
    b += product
    
    # Formatted row
    print(f"{i+1:<4} | {a_coefficient[i]:<12.4f} | {high_val:<15.2f} | {low_val:<10.2f} | {diff:<12.4f} | {product:<10.4f} | {b:.4f}")

# Calculate Final W
sample_mean = sum(test_data) / N
ss = sum([(x - sample_mean)**2 for x in test_data])
w_statistic = (b**2) / ss

print("-" * 90)
print(f"Final Sum of Products (b): {b:.4f}")
print(f"Sum of Squares (SS): {ss:.4f}")
print(f"Calculated W-Statistic: {w_statistic:.4f}")

i    | coeff (a_i)  | High (x_n-i+1)  | Low (x_i)  | Difference   | Product    | Running b
------------------------------------------------------------------------------------------
1    | 0.3751       | 249.24          | 229.35     | 19.8900      | 7.4607     | 7.4607
2    | 0.2574       | 246.63          | 234.34     | 12.2900      | 3.1634     | 10.6242
3    | 0.2260       | 245.50          | 235.36     | 10.1400      | 2.2916     | 12.9158
4    | 0.2032       | 245.22          | 235.53     | 9.6900       | 1.9690     | 14.8848
5    | 0.1847       | 244.92          | 235.72     | 9.2000       | 1.6992     | 16.5841
6    | 0.1691       | 244.82          | 235.91     | 8.9100       | 1.5067     | 18.0908
7    | 0.1554       | 244.52          | 236.79     | 7.7300       | 1.2012     | 19.2920
8    | 0.1430       | 244.36          | 237.02     | 7.3400       | 1.0496     | 20.3416
9    | 0.1317       | 243.95          | 237.45     | 6.5000       | 0.8561     | 21.1977
10   | 0.1212     

# Calculation of p value done using [table](https://real-statistics.com/wp-content/uploads/2012/12/image3742.png) reference

In [13]:
# Formula to calculate p value
# p = p1 + ((p2-p1)/(w2-w1))*(w-w1)
# Where w1 is the largest value in the row that is just less than w while W2 is the smallest value in the same row that is just greater than w
# p1 and p2 are corresponding values of w1 and w2 respectively

# we get
w1 = 0.974
w2 = 0.988
p1 = 0.5 
p2 = 0.95
w = 0.985
# then

p = p1 + ((p2-p1)/(w2-w1))*(w-w1)
print("P value = ", p)

P value =  0.8535714285714285


**Findings**

The shapiro() function gave us $0.9812$, while the manual loop gave us $0.9856$. This happened because of the coefficients ($a_i$):
- The Table Values: The values like $0.3751$ we used are rounded constants from standard statistical tables. These tables are designed for hand-calculations.
- The Python Values: The scipy.stats.shapiro function uses the Royston algorithm, which calculates coefficients to a much higher precision (often 8+ decimal places) and uses a slightly different weight for the very last terms.
- The Impact: Because $W = b^2/SS$, even a tiny change in $b$ is squared, which amplifies the error. A difference of $0.004$ in $W$ might seem small, but in a Shapiro-Wilk table, that covers a massive jump in $p$-value.

The relationship between $W$ and the $p$-value is not a straight line; it is a curve
- Linear Interpolation we used: This assumes that the jump from $W_1$ to $W_2$ is a perfectly straight line.
- The Reality: Between $p=0.50$ and $p=0.95$, the curve flattens out significantly.

### The Royston Method
For a sample size of n=50, we transform $W$ into a value $y$, then into a $Z$-score.

In [14]:
# Calculation of y
# y = ln(1-w)

import math

y = math.log(1-w)
print(y)

-4.199705077879926


Calculate the Expected Mean ($\mu$) and Standard Deviation ($\sigma$)

For sample sizes $12 \le n \le 2000$, Royston provided this formula using the natural log of $n$, 
which we'll call $L = \ln(n)$:$$\mu = -1.5861 - 0.31082L - 0.083751L^2 + 0.0038915L^3$$
and $$\sigma = e^{( -0.4803 - 0.082676L + 0.0030302L^2 )}$$


In [15]:
# For n=50, we have
l = math.log(50)
l2 = l*l
l3 = l*l*l

print(l,l2,l3)

3.912023005428146 15.303923994999064 59.86930274176016


In [16]:
# Therefore, meu becomes
meu = -1.5861 - 0.31082*l - 0.083751*l2 + 0.0038915*l3
print("meu = ",meu)

meu =  -3.8507725374327837


In [17]:
# Ans, standard deviation becomes
exponent = -0.4803 - (0.082676 * l) + (0.0030302 *l*l)
sigma = math.exp(exponent)
print("sigma = ",sigma)

sigma =  0.46890435581026263


In [18]:
# Now, getting the z valuy

z = (y-meu)/sigma
print("z score = ",z)

z score =  -0.7441443785357682


[Z tale](https://cdn.prod.website-files.com/6634a8f8dd9b2a63c9e6be83/669a129895996d732eec4a11_451912.image0.jpeg) Lookup

For z score of -0.74, we have area coverage = 0.296

In [19]:
area_coverage = 0.296
p = 1 - area_coverage # since we want the upper tail (Represents that the adata fits normal pattern)

print("P value using royston method = ", p)

P value using royston method =  0.704


**Interpretation:**

Because the $p$-value is significantly higher than the threshold, we fail to reject the Null Hypothesis

### Corelation and Covarience of `Global_active_power` and `Global_intensity`

In [20]:
test_df = df[['Global_active_power', 'Global_intensity']]
test_df.cov()

Global_active_power  Global_intensity
Global_active_power             1.117871          4.693812
Global_intensity                4.693812         19.752658

In [21]:
test_df.corr()

Global_active_power  Global_intensity
Global_active_power             1.000000          0.998889
Global_intensity                0.998889          1.000000

**Interpretation**

The covariance between the test variables `Global_active_power` and `Global_intensity` is 4.69 which is positive indicating they move together (due to same sign)

The correlation  between them is 0.998 indicates that they have a very strong positive linear relationship. When Global_active_power increases, Global_intensity increases almost proportionally. One variable can predict the other very well

### Step wise calculation of corelation and covarience

**Coverience**
$$cov(X, Y) = \frac{\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}{n - 1}$$

In [43]:
x_bar = test_df['Global_active_power'].mean()
y_bar = test_df['Global_intensity'].mean()
n = len(test_df['Global_active_power'])

deviation_of_X = test_df['Global_active_power'] - x_bar
deviation_of_y = test_df['Global_intensity'] - y_bar

product_of_deviation = deviation_of_X * deviation_of_y
aggrigate = product_of_deviation.sum()

coverience = aggrigate/(n-1)

print("Coverience between Global_active_power and Global_intensity = ",coverience)

Coverience between Global_active_power and Global_intensity =  4.693811709418095


**Corelation**
$$r = \frac{\sum (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum (x_i - \bar{x})^2 \sum (y_i - \bar{y})^2}}$$

OR

$$r = \frac{cov(X, Y)}{s_x s_y}$$

In [60]:
aggrigate_deviation_of_X_squared = (deviation_of_X * deviation_of_X).sum()
aggrigate_deviation_of_y_squared = (deviation_of_y * deviation_of_y).sum()

product_of_deviation_squared = aggrigate_deviation_of_X_squared * aggrigate_deviation_of_y_squared
sx_sy = math.sqrt(product_of_deviation_squared)
sx_sy

# finally
r = aggrigate/sx_sy
print("Computed corelation = ", r)

Computed corelation =  0.9988886002095894
